server/src/inmantals/server.py

```diff
@@ -64,9 +65,149 @@ def __init__(self, message: str) -> None:
+    """
+
+    folder_uri: URI
+    name: str
+    inmanta_project_dir: Union[Optional[tempfile.TemporaryDirectory], str]
+    handler: "InmantaLSHandler"
+
+    def __init__(self, folder_uri: URI, name: str, handler: "InmantaLSHandler"):
+        self.folder_uri = folder_uri
+        self.name = name
+        self.handler = handler  # Keep a reference to the handler for cleanup
+
+        # Check that we are working inside an existing project:
+        project_file: str = os.path.join(folder_uri.path, module.Project.PROJECT_FILE)
+        if os.path.exists(project_file):
+            self.inmanta_project_dir = os.path.dirname(project_file)
+        else:
+            self.inmanta_project_dir = None
+
+    def get_project_dir(self) -> Optional[str]:
+        """
+        If this folder holds an inmanta project, this returns the path to this project.
+        If this folder holds an inmanta module, this returns the path of the temporary
+        inmanta project used to load this module.
+        """
+        if not self.inmanta_project_dir:
+            return None
+        if isinstance(self.inmanta_project_dir, str):
+            return self.inmanta_project_dir
+        else:
+            return self.inmanta_project_dir.name
+
+    def get_setting(self, key: str, default: Optional[str] = None) -> Optional[str]:
+        """
+        Get an inmanta extension setting for this folder
+        """
+        if key == "compiler_venv_path":  # support for versions of core that require a compiler venv.
+            return os.path.join(self.get_folder_path(), ".env-ls-compiler")
+        try:
+            # Look for some custom settings for this folder
+            with open(os.path.join(self.get_folder_path(), ".vscode", "settings.json")) as settings_file:
+                settings: dict[str, str] = json.load(settings_file)
+                return settings.get(key, default)
+        except FileNotFoundError:
+            return default
+
+    @classmethod
+    def unpack_workspaces(cls, workspace_folders: Sequence[object], ls_handler: "InmantaLSHandler") -> Dict[str, "Folder"]:
+        return {folder["uri"]: Folder(URI(folder["uri"]), folder["name"], ls_handler) for folder in workspace_folders}
+
+    def cleanup(self) -> None:
+        logger.info(f"calling cleanup for {self}")
+        if self.inmanta_project_dir:
+            self.handler.remove_folder(str(self.folder_uri))
+            self.inmanta_project_dir.cleanup()
+
+    def get_folder_path(self) -> str:
+        """
+        Convert the folders's uri into a regular path
+        (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#uri)
+        """
+        return self.folder_uri.path
+
+    def create_tmp_project(self) -> str:
+        """
+        When working on a module, an inmanta project is created in a temporary directory.
+        """
+        self.inmanta_project_dir = tempfile.TemporaryDirectory()
```

Comment:
Do I understand correctly that the caller of `create_tmp_project` is expected to call `cleanup` at some later point? If so, it may be worth adding that invariant to the docstring. Could it also be useful to add a context manager that does the cleanup automatically?
```python
with Folder.tmp_project() as something:
    do_stuff(something)  # doesn't matter if this raises an exception because the context manager cleans up in a `finally` block
```

Reply:
> Do I understand correctly that the caller of create_tmp_project is expected to call cleanup at some later point? 
Yes exactly. Here's how I see the flow:
When the server first realizes the open folder is a module -> create a tmp project that we should keep around until server shutdown
When a .cf file of the module is saved -> use the existing tmp project.

I considered using a context manager for this and use it inside the compile_and_anchor method but decided against it because it would mean  creating a new tmp project and tearing it down everytime this method is called. I figured it would be better to reuse the project created the first time around.

Reply:
Ok, I have no objections against that. If not too much trouble it would be good to trigger the cleanup also when the server is killed or crashes. I don't know how much effort that would be.